### DISCUSSION TITLE
How to use a second path to display data?

### DISCUSSION THREAD

**Discussion by vfresnaud** [pymeasure/pymeasure#840] [Q&A] [answered]
Hello,
I'm using the GUI version of the experiment.
I've seen that in the procedure we are able to emit on the topic 'results' and 'progress'. 
The usage of 'results' is convenient to store and display data. I would like to emit on another topic to plot temporary data that I do not want to store.
I was expecting to start a second thread that will be listening to the new topic and will be able to display its own data, but i do not find a proper way to do it. 
Is there an example on how to add a listener, or add to the queue a thread that will process data that we do not want to save? Or on a dedicated widget that can consume data other that the ones coming from the 'results' emission?
I do not see where the listener needs to be instantiate (I have a zmq context issue)

in my `class CustomMainWindow(ManagedWindow):`
i've tried to add a listener when i'm a queuing the experiment (i've assumed the worker is started once the experiment is started)

```
    def queue(self, *, procedure=None):
        l_filename = unique_filename(test_ctx.bench_sect_dict['TEST_CONFIGURATION']['test_result_folder_path'], test_ctx.bench_sect_dict['TEST_CONFIGURATION']['webapp_test_def'])
        if procedure is None:
            procedure = self.make_procedure()
        print("DEBUG: " + test_ctx.bench_sect_dict['TEST_CONFIGURATION']['test_result_folder_path'])

        proc_results = Results(procedure, l_filename)
        experiment = self.new_experiment(proc_results)
        self.manager.queue(experiment)
        capture = pymeasure.experiment.listeners.Listener(port=5888, topic='fft', timeout=0.01)
        capture.start()
```
but it ends in the following issue:
```
File "\venv\lib\site-packages\pymeasure\experiment\listeners.py", line 68, in __init__
    self.context = zmq.Context()
AttributeError: 'NoneType' object has no attribute 'Context'
```
i think zmq is not properly handled... but it is not clear to me how it should be started...

If someone can point me where to look at, or give some hints, it will help :)
 
Cheers,

Vincent

**Comment by bilderbuchi** [comment, marked as answer]
It could be you don't have `pyzmq` and `cloudpickle` installed?

**Comment by vfresnaud** [comment]
.... ouaw... indeed it solved the issue wit the error message.. i was thinking that by default all the Worker/Results was already using ZMQ and that it was installed with pymeasure. but i was wrong indeed.  
Now, i will dig to figure out on how to use it.

thank you for your prompt reply!

**Reply by bilderbuchi** [reply]
We have it as an optional dependency, so people not needing don't get it for no good reason. :-)
